### PR TITLE
SSL tests: send some application data.

### PR DESCRIPTION
**Pull request by ekasper**

### CODE REVIEW

File: test/handshake_helper.c [hunk 1]
```diff
@@ -401,13 +401,131 @@ static void configure_handshake_ssl(SSL *server, SSL *client,
...
-
+/* The status for each connection phase. */
 typedef enum {
     PEER_SUCCESS,
     PEER_RETRY,
     PEER_ERROR
 } peer_status_t;
 
+/* An SSL object and associated read-write buffers. */
+typedef struct {
```
Comment (richsalz) [inline test/handshake_helper.c]: okay, here i'm on more solid ground.  :)  456 of 571 have a name for the struct.

File: test/ssl_test_ctx_test.c [hunk 2]
```diff
@@ -240,6 +245,7 @@ static int test_good_configuration()
     fixture.test_section = "ssltest_good";
     fixture.expected_ctx->method = SSL_TEST_METHOD_DTLS;
     fixture.expected_ctx->handshake_mode = SSL_TEST_HANDSHAKE_RESUME;
+    fixture.expected_ctx->app_data_size = 1024;
```
Comment (richsalz) [inline test/ssl_test_ctx_test.c]: isn't this specified in the conf file?

File: test/handshake_helper.c [hunk 3]
```diff
@@ -401,13 +401,131 @@ static void configure_handshake_ssl(SSL *server, SSL *client,
...
+
+    /* We read everything available... */
+    while (ret > 0 && peer->bytes_to_read) {
+        ret = SSL_read(peer->ssl, peer->read_buf, peer->read_buf_len);
+        if (ret > 0) {
+            TEST_check(ret <= peer->bytes_to_read);
+            peer->bytes_to_read -= ret;
+        } else if (ret == 0) {
+            peer->status = PEER_ERROR;
+            return;
```
Comment (richsalz) [inline test/handshake_helper.c]: i have else after return but not enough to be a jerk about it here.

File: test/handshake_helper.c [hunk 4]
```diff
@@ -614,23 +773,24 @@ static HANDSHAKE_RESULT *do_handshake_internal(
...
-            if (shutdown) {
+            phase = next_phase(phase);
+            if (phase == CONNECTION_DONE) {
                 ret->result = SSL_TEST_SUCCESS;
                 goto err;
             } else {
-                client_status = server_status = PEER_RETRY;
-                shutdown = 1;
+                client.status = server.status = PEER_RETRY;
+                /* For now, client starts each phase. */
```
Comment (richsalz) [inline test/handshake_helper.c]: does this deserve an XXX or TODO note?

File: test/handshake_helper.c [hunk 5]
```diff
@@ -401,13 +401,131 @@ static void configure_handshake_ssl(SSL *server, SSL *client,
...
+    unsigned char *read_buf;
+    int read_buf_len;
+    int bytes_to_write;
+    int bytes_to_read;
+    peer_status_t status;
+} PEER;
+
+static void create_peer(PEER *peer, SSL_CTX *ctx)
+{
+    static const int peer_buffer_size = 64 * 1024;
```
Comment (richsalz) [inline test/handshake_helper.c]: blank line.  also, should sanity-check that the conf file doesn't exceed this number?

File: test/ssl_test_ctx.c [hunk 6]
```diff
@@ -133,7 +168,7 @@ const char *ssl_protocol_name(int protocol)
 }
 
 /***********************/
-/* VerifyCallback. */
+/* VerifyCallback.     */
```
Comment (richsalz) [inline test/ssl_test_ctx.c]: ah, the tyranny of block comments.  yuk :)

File: test/ssl_test_ctx.c [hunk 7]
```diff
@@ -439,6 +453,7 @@ SSL_TEST_CTX *SSL_TEST_CTX_new()
     SSL_TEST_CTX *ret;
     ret = OPENSSL_zalloc(sizeof(*ret));
     TEST_check(ret != NULL);
+    ret->app_data_size = 256;
```
Comment (richsalz) [inline test/ssl_test_ctx.c]: make a file-global constant or define for this number.

**Comment by richsalz** [comment]
I know I'm being dumb, but where is the actual app data that is sent?

File: test/ssl_test_ctx_test.c (hunk 2 above)
Comment (ekasper) [inline test/ssl_test_ctx_test.c]: That's the test expectation that says we should be getting 1024 from the conf.

File: test/handshake_helper.c (hunk 1 above)
Comment (ekasper) [inline test/handshake_helper.c]: I've no idea what it's good for but I've added a name :)

File: test/handshake_helper.c (hunk 4 above)
Comment (ekasper) [inline test/handshake_helper.c]: Expanded on the comment.

File: test/handshake_helper.c (hunk 5 above)
Comment (ekasper) [inline test/handshake_helper.c]: Nope, that's just the amount of data to send in one go. I made this up (the old tests use 40kB) and will tweak this after I understand how the multi-buffer tests work.

File: test/ssl_test_ctx.c (hunk 7 above)
Comment (ekasper) [inline test/ssl_test_ctx.c]: Done, constant.

**Comment by ekasper** [comment]
@richsalz - we just send the nul-bytes from the OPENSSL_zalloc'ed `write_buf` over and over again. I might tweak this to send some specific content later, so that we can also test that we get the correct bytes (not just the correct amount of bytes).

**Comment by richsalz** [comment]
okay, you've addressed all my issues.  +1

**Pull request closed without merging**